### PR TITLE
Support typescript 3.6

### PR DESCRIPTION
**Pull request by fisker**
- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [ ] (If the change is user-facing) I’ve added my changes to the `CHANGELOG.unreleased.md` file following the template.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by evilebottnawi** [comment]
What is status?

**Comment by fisker** [comment]
There is a  problem I can't fix

```typescript
type A = & string;
```
this code will format into

```typescript
type A = string;
```
the previous snapshot is also the same  but new version parser parse different ast cause failure test

I tried to update the snap to keep the ast match, but I can't figure a  way to detect

**Comment by lydell** [comment]
What does `type A = & string;` mean in TypeScript?

**Comment by fisker** [comment]
I'm not be familiar with Typescripts types, after a few search, I believe

```typescript
type A = & string;
```

is not equal to

```typescript
type A = string;
```

with `&` leading operator, it's Intersection Type, otherwise it is not, this is the reason AST compare fails, if this is currect, I might fix a possible bug.

We might want some one familiar with this to review

**Comment by fisker** [comment]
why CI not running?

**Comment by lydell** [comment]
Wouldn’t it be nicer if we cleaned `& foo` (and `| foo`) to just `foo`? I get that since the TypeScript update it technically changes the AST, but we already have small deviations when it makes sense (such as for JSX). You should be able to make the tests pass by updating `src/language-js/clean.js`. It already has a couple of TypeScript cases.

**Comment by fisker** [comment]
@lydell 

I really don't know those things, but according to the spec `& foo` `| foo` is different type with `foo`, so we might also add `| foo` test, and keep the leading `|`, maybe after this line https://github.com/prettier/prettier/blob/fced57c1d68971407481247f35fd2c6b329fce4e/tests/typescript/conformance/types/union/unionTypeEquivalence.ts#L4

**Comment by fisker** [comment]
I mean `foo` `& foo` `| foo` has different AST, let me try to clean

**Comment by fisker** [comment]
Good to go

File: src/language-js/clean.js [hunk 1]
```diff
@@ -65,6 +65,14 @@ function clean(ast, newObj, parent) {
     return newObj.typeAnnotation;
   }
 
+  // (TypeScript) `& foo` and `| foo` into `foo`
```
Comment (lydell) [inline src/language-js/clean.js]:
```suggestion
  // (TypeScript) Bypass `& foo` and `| foo` into `foo`
  // https://github.com/microsoft/TypeScript/issues/30995
```

File: scripts/build/config.js [hunk 2]
```diff
@@ -42,6 +42,12 @@ const parsers = [
     replace: {
       // node v4 compatibility for @typescript-eslint/typescript-estree
       "(!unique.includes(raw))": "(unique.indexOf(raw) === -1)"
+    },
+    commonjs: {
+      ignore: [
+        // optional typescript etw logger
```
Comment (lydell) [inline scripts/build/config.js]:
```suggestion
        // Optional package for TypeScript that logs ETW events (a Windows-only technology).
```

File: package.json [hunk 3]
```diff
@@ -129,7 +129,7 @@
     "lint": "cross-env EFF_NO_LINK_RULES=true eslint . --format friendly",
     "lint-docs": "prettylint {.,docs,website,website/blog}/*.md",
     "lint-dist": "eslint --no-eslintrc --no-ignore --env=browser \"dist/!(bin-prettier|index|third-party).js\"",
-    "build": "node --max-old-space-size=2048 ./scripts/build/build.js",
+    "build": "node --max-old-space-size=3072 ./scripts/build/build.js",
```
Comment (lydell) [inline package.json]: Does the build fail without increasing this?

Comment (fisker) [reply package.json]: yes, not enough memory

Comment (fisker) [reply package.json]: see https://dev.azure.com/prettier/prettier/_build/results?buildId=1672

Comment (lydell) [reply package.json]: Ouch. I wonder what’s taking so much memory…

**Pull request merged**